Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -89,24 +89,37 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
     clearTimebase();
 }
 
+size_t VideoMediaSampleRenderer::decodedSamplesCount() const
+{
+    return m_decodedSampleQueue ? PAL::CMBufferQueueGetBufferCount(m_decodedSampleQueue.get()) : 0;
+}
+
 bool VideoMediaSampleRenderer::isReadyForMoreMediaData() const
 {
     assertIsMainThread();
 
-    if (m_framesBeingDecoded >= CompressedSampleQueueHighWaterMark)
+    if (m_framesBeingDecoded + decodedSamplesCount() >= SampleQueueHighWaterMark)
         return false;
 
     return [renderer() isReadyForMoreMediaData];
 }
 
+void VideoMediaSampleRenderer::queueMaybeBecomeReadyForMoreMediaData()
+{
+    callOnMainThread([weakThis = ThreadSafeWeakPtr { *this }] {
+        if (RefPtr protectedThis = weakThis.get())
+            protectedThis->maybeBecomeReadyForMoreMediaData();
+    });
+}
+
 void VideoMediaSampleRenderer::maybeBecomeReadyForMoreMediaData()
 {
     assertIsMainThread();
 
     if (![renderer() isReadyForMoreMediaData])
         return;
 
-    if (m_framesBeingDecoded >= CompressedSampleQueueLowWaterMark)
+    if (m_framesBeingDecoded + decodedSamplesCount() >= SampleQueueLowWaterMark)
```

Comment:
This test is in two places, it would be nice to have a utility function.

Reply:
they aren't quite the same. One is using the LowWaterMark and the other the HighWaterMark